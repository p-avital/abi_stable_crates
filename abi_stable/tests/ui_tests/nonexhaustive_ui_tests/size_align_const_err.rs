use abi_stable::{
    nonexhaustive_enum::NonExhaustiveFor as NEFor,
    StableAbi,
};

#[repr(u8)]
#[derive(StableAbi)]
#[sabi(kind(WithNonExhaustive(
    size = 1,
    align = 1,
)))]
#[sabi(with_constructor)]
pub enum TooLarge<T = u8> {
    Foo,
    Bar,
    Baz(T),
}

const _: () = { std::mem::forget(NEFor::new(<TooLarge>::Foo)); };


#[repr(u8)]
#[derive(StableAbi)]
#[sabi(kind(WithNonExhaustive(
    size = 32,
    align = 1,
)))]
#[sabi(with_constructor)]
pub enum Unaligned<T = u64> {
    Foo,
    Bar,
    Baz(T),
}

const _: () = { std::mem::forget(NEFor::new(<Unaligned>::Foo)); };

#[repr(u8)]
#[derive(StableAbi)]
#[sabi(kind(WithNonExhaustive(
    size = {one()},
    align = {one()},
)))]
#[sabi(with_constructor)]
pub enum UnalignedAndTooLarge<T = u64> {
    Foo,
    Bar,
    Baz(T),
}

const _: () = { std::mem::forget(NEFor::new(<UnalignedAndTooLarge>::Foo)); };

const fn one() -> usize {
    1
}


fn main(){}