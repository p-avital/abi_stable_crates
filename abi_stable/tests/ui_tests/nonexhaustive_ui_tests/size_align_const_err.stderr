error[E0080]: evaluation of constant value failed
 --> src/inline_storage.rs
  |
  |         Self::assert_fits_within_storage();
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |         |
  |         the evaluated program panicked at '
The size of the storage is smaller than the contained type:
enum_: "<unknown>"
enum_size: 2
enum_alignment: 1
storage_: "<unknown>"
storage_size: 1
storage_alignment: 1
', $DIR/src/inline_storage.rs:193:9
    |         inside `inline_storage::ScratchSpace::<TooLarge, TooLarge_Storage>::new` at $DIR/src/inline_storage.rs:193:9
    |
   ::: src/nonexhaustive_enum/nonexhaustive.rs
    |
    |         NonExhaustive::with_storage_and_interface(value)
    |         ------------------------------------------------ inside `NonExhaustive::<TooLarge, TooLarge_Storage, ()>::new` at $DIR/src/nonexhaustive_enum/nonexhaustive.rs:231:9
...
    |         unsafe { NonExhaustive::with_vtable(value, E::VTABLE) }
    |                  -------------------------------------------- inside `NonExhaustive::<TooLarge, TooLarge_Storage, ()>::with_storage_and_interface` at $DIR/src/nonexhaustive_enum/nonexhaustive.rs:272:18
    |
   ::: tests/ui_tests/nonexhaustive_ui_tests/size_align_const_err.rs:19:34
    |
19  | const _: () = { std::mem::forget(NEFor::new(<TooLarge>::Foo)); };
    |                                  --------------------------- inside `_` at $DIR/tests/ui_tests/nonexhaustive_ui_tests/size_align_const_err.rs:19:34

error[E0080]: evaluation of constant value failed
 --> src/inline_storage.rs
  |
  |         Self::assert_fits_within_storage();
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |         |
  |         the evaluated program panicked at '
The alignment of the storage is lower than the contained type:
enum_: "<unknown>"
enum_size: 16
enum_alignment: 8
storage_: "<unknown>"
storage_size: 32
storage_alignment: 1
', $DIR/src/inline_storage.rs:193:9
    |         inside `inline_storage::ScratchSpace::<Unaligned, Unaligned_Storage>::new` at $DIR/src/inline_storage.rs:193:9
    |
   ::: src/nonexhaustive_enum/nonexhaustive.rs
    |
    |         NonExhaustive::with_storage_and_interface(value)
    |         ------------------------------------------------ inside `NonExhaustive::<Unaligned, Unaligned_Storage, ()>::new` at $DIR/src/nonexhaustive_enum/nonexhaustive.rs:231:9
...
    |         unsafe { NonExhaustive::with_vtable(value, E::VTABLE) }
    |                  -------------------------------------------- inside `NonExhaustive::<Unaligned, Unaligned_Storage, ()>::with_storage_and_interface` at $DIR/src/nonexhaustive_enum/nonexhaustive.rs:272:18
    |
   ::: tests/ui_tests/nonexhaustive_ui_tests/size_align_const_err.rs:35:34
    |
35  | const _: () = { std::mem::forget(NEFor::new(<Unaligned>::Foo)); };
    |                                  ---------------------------- inside `_` at $DIR/tests/ui_tests/nonexhaustive_ui_tests/size_align_const_err.rs:35:34

error[E0080]: evaluation of constant value failed
 --> src/inline_storage.rs
  |
  |         Self::assert_fits_within_storage();
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |         |
  |         the evaluated program panicked at '
The alignment and size of the storage is smaller than the contained type:
enum_: "<unknown>"
enum_size: 16
enum_alignment: 8
storage_: "<unknown>"
storage_size: 1
storage_alignment: 1
', $DIR/src/inline_storage.rs:193:9
    |         inside `inline_storage::ScratchSpace::<UnalignedAndTooLarge, UnalignedAndTooLarge_Storage>::new` at $DIR/src/inline_storage.rs:193:9
    |
   ::: src/nonexhaustive_enum/nonexhaustive.rs
    |
    |         NonExhaustive::with_storage_and_interface(value)
    |         ------------------------------------------------ inside `NonExhaustive::<UnalignedAndTooLarge, UnalignedAndTooLarge_Storage, ()>::new` at $DIR/src/nonexhaustive_enum/nonexhaustive.rs:231:9
...
    |         unsafe { NonExhaustive::with_vtable(value, E::VTABLE) }
    |                  -------------------------------------------- inside `NonExhaustive::<UnalignedAndTooLarge, UnalignedAndTooLarge_Storage, ()>::with_storage_and_interface` at $DIR/src/nonexhaustive_enum/nonexhaustive.rs:272:18
    |
   ::: tests/ui_tests/nonexhaustive_ui_tests/size_align_const_err.rs:50:34
    |
50  | const _: () = { std::mem::forget(NEFor::new(<UnalignedAndTooLarge>::Foo)); };
    |                                  --------------------------------------- inside `_` at $DIR/tests/ui_tests/nonexhaustive_ui_tests/size_align_const_err.rs:50:34
